content/blog/ggplot-3-5-0-guides/index.Rmd

```diff
@@ -0,0 +1,887 @@
+#|  clipped."
+
+p + aes(fill = model)
+```
+
+Another issue that may come up is that the 'available space' might be 0. 
+Because the plot itself is also space-filling, setting null-heights for top/bottom positions or null-widths for left/right positions means there is no available space.
+This may result in the keys or bars becoming invisible.
+For the plot below, recall that we've set the `legend.key.height` setting to a null unit.
+
+```{r disappearing_keys}
+#| fig.alt = "Still the same scatterplot but without the fill variable. Legends
+#|  are placed at the top of the panel, but the bar and key backgrounds have
+#|  disappeared. The text labels are still present."
+
+p + theme(legend.position = "top")
+```
+
+### Other improvements
+
+We welcome a new type of legend: `guide_custom()`. 
+It can be used to add any graphical object (grob) to a plot, like `annotation_custom()`. 
+There are a few differences though: it is positioned just like a legend and adds titles and margins.
+In some sense, this guide is 'special', as it is the only guide that does not directly reflect a scale.
+The downside is that it cannot read properties from the plot, but the upside is that it is very flexible.
+Be careful that if your grob does not have an absolute size, to set the `width` and `height` arguments.
+
+```{r custom_guide}
+#| fig.alt = "A map of the US state North Carolina, where fill colour indicates
+#|  the area of counties. Underneath the colour bar for the fill, there is an
+#|  eight-pointed star to the right of the panel with the title 'compass'."
+
+x <- c(0.5, 1, 1.5, 1.2, 1.5, 1, 0.5, 0.8, 1, 1.15, 2, 1.15, 1, 0.85, 0, 0.85)
+y <- c(1.5, 1.2, 1.5, 1, 0.5, 0.8, 0.5, 1, 2, 1.15, 1, 0.85, 0, 0.85, 1, 1.15)
+
+compass_rose <- grid::polygonGrob(
+  x = unit(x, "cm"), y = unit(y, "cm"), id.lengths = c(8, 8),
+  gp = grid::gpar(fill = c("grey50", "grey25"), col = NA)
+)
+
+nc <- sf::st_read(system.file("shape/nc.shp", package = "sf"), quiet = TRUE)
+ggplot(nc) +
+  geom_sf(aes(fill = AREA)) +
+  guides(custom = guide_custom(compass_rose, title = "compass"))
+```
+
+In previous version of ggplot2, when legend titles are wider than the legends, the guide-title alignment was always left aligned.
+Now, the justification setting of the legend text determines the alignment: 1 is right or top aligned and 0 is left or bottom aligned.
+
+```{r title_justification}
+#| fig.alt = "Scatterplot of engine displacement versus highway miles per 
+#|  gallon. The 'drv' variable has a legend that is left aligned, whereas the
+#|  'cyl' variable has a legend that is right-aligned."
+
+ggplot(mpg, aes(displ, hwy, shape = factor(cyl), colour = drv)) +
+  geom_point() +
+  guides(
+    shape = guide_legend(
+      title = "A title that is pretty long",
+      theme = theme(legend.title = element_text(hjust = 1))
+    ),
+    colour = guide_legend(
+      title = "Another long title",
+      theme = theme(legend.title = element_text(hjust = 0))
+    )
+  )
+```
+
+## Axes
```

Comment:
This blog post is really long and I got tired of reading/reviewing around here. I think that's a sign that this would be better as 3 separate posts covering, legends, axes, and extensions.

Reply:
Will do!